File: Concorrente/Java/virtualthread/src/main/java/com/imd/InversoDistanciaPonderada.java
package com.imd;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class InversoDistanciaPonderada {

    public Ponto atualizarVizinhosMaisProximos(Ponto alvo,
            List<Ponto> candidatos,
            int k) {
        PriorityQueue<Ponto> heap = new PriorityQueue<>(k,
                (p1, p2) -> {
                    double d1 = distanciaGeografica(alvo.getLatitude(), alvo.getLongitude(),
                            p1.getLatitude(), p1.getLongitude());
                    double d2 = distanciaGeografica(alvo.getLatitude(), alvo.getLongitude(),
                            p2.getLatitude(), p2.getLongitude());
                    return Double.compare(d2, d1);
                });

        for (Ponto candidato : candidatos) {
            if (heap.size() < k) {
                heap.offer(candidato);
            } else {
                Ponto pior = heap.peek();
                double dPior = distanciaGeografica(alvo.getLatitude(), alvo.getLongitude(),
                        pior.getLatitude(), pior.getLongitude());
                double dCand = distanciaGeografica(alvo.getLatitude(), alvo.getLongitude(),
                        candidato.getLatitude(), candidato.getLongitude());
                if (dCand < dPior) {
                    heap.poll();
                    heap.offer(candidato);
                }
            }
        }
        List<Ponto> vizinhosK = new ArrayList<>(heap);

        alvo.setPontosProximos(vizinhosK);
        return alvo;
    }

    public double interpolarComVizinhos(Ponto alvo, int p) {
        double eps = 1e-9;
        double somaIguais = 0.0;
        int contIguais = 0;

        for (Ponto viz : alvo.getPontosProximos()) {
            double d = distanciaGeografica(
                    alvo.getLatitude(), alvo.getLongitude(),
                    viz.getLatitude(), viz.getLongitude());
            if (d < eps) {
                somaIguais += viz.getTemperatura();
                contIguais++;
            }
        }

        if (contIguais > 0) {
            return somaIguais / contIguais;
        }

        double numerador = 0.0;
        double denominador = 0.0;
        for (Ponto viz : alvo.getPontosProximos()) {
            double d = distanciaGeografica(
                    alvo.getLatitude(), alvo.getLongitude(),
                    viz.getLatitude(), viz.getLongitude());
            double peso = 1.0 / Math.pow(d, p);
            numerador += peso * viz.getTemperatura();
            denominador += peso;
        }
        return numerador / denominador;
    }

    public static double distanciaGeografica(double lat1, double lon1,
            double lat2, double lon2) {
        double R = 6371.0;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                        * Math.sin(dLon / 2) * Math.sin(dLon / 2);
        return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    }
}

File: Concorrente/Elixir/idw/lib/inverso_distancia_ponderada.ex
defmodule Idw.InversoDistanciaPonderada do
  alias Idw.Ponto
  @raio_terra_km 6371.0
  def atualizar_vizinhos_mais_proximos(alvo, candidatos, k) do
    heap =
      Enum.reduce(candidatos, [], fn candidato, acc ->
        acc = [candidato | acc]

        if length(acc) > k do
          Enum.sort_by(acc, fn p ->
            -distancia_geografica(alvo.latitude, alvo.longitude, p.latitude, p.longitude)
          end)
          |> Enum.take(k)
        else
          acc
        end
      end)

    %Ponto{alvo | pontos_proximos: heap}
  end

  def interpolar_com_vizinhos(alvo, p) do
    eps = 1.0e-9

    iguais =
      Enum.filter(alvo.pontos_proximos, fn viz ->
        distancia_geografica(alvo.latitude, alvo.longitude, viz.latitude, viz.longitude) < eps
      end)

    if length(iguais) > 0 do
      Enum.map(iguais, & &1.temperatura)
      |> Enum.sum()
      |> Kernel./(length(iguais))
    else
      Enum.reduce(alvo.pontos_proximos, {0.0, 0.0}, fn viz, {num, den} ->
        d = distancia_geografica(alvo.latitude, alvo.longitude, viz.latitude, viz.longitude)
        peso = 1.0 / :math.pow(d, p)
        {num + peso * viz.temperatura, den + peso}
      end)
      |> (fn {num, den} -> num / den end).()
    end
  end

  def distancia_geografica(lat1, lon1, lat2, lon2) do
    r = 6371.0
    dlat = :math.pi() * (lat2 - lat1) / 180.0
    dlon = :math.pi() * (lon2 - lon1) / 180.0

    a =
      :math.sin(dlat / 2) * :math.sin(dlat / 2) +
        :math.cos(:math.pi() * lat1 / 180.0) * :math.cos(:math.pi() * lat2 / 180.0) *
          :math.sin(dlon / 2) * :math.sin(dlon / 2)

    r * 2 * :math.atan2(:math.sqrt(a), :math.sqrt(1 - a))
  end
end
